### PR TITLE
[WIP] Implement Lookahead optimizer

### PR DESCRIPTION
**Pull request by hvy**
Partially fixes https://github.com/chainer/chainer/issues/8153, implementing the lookahead optimizer (https://arxiv.org/abs/1907.08610).

This PR mostly aims to demonstrate a possible implementation and might not be suitable in Chainer in which case we can close this PR (this optimizer might not be suitable at all considering the fact that PyTorch rejected it). 

It introduces a new `GradientMethod ` that is slightly different from the existing ones in that it accepts an `Optimizer` instance as an argument and updates using that one in the "inner loop". ~Thus, there are some tricks in order to get around initialization/serialization/deserialization such as overriding `setup` and `_init_states`. I might be missing some.~ There are other ways to implement this optimizer e.g. using arguments to existing optimizers or in the `Updater` layer but I didn't think those would be easier to maintain. I'd be more than happy to hear other ideas.

### TODO
- [ ] Write tests.
- [ ] Write docs.
- [ ] Compare results with paper.

### CODE REVIEW

**Comment by hvy** [comment]
Overwriting the update rules in setup we cannot do of course. I now feel nesting optimizers isn’t really ideal.

**Comment by hvy** [comment]
Changed to instead patch the given base optimizer update rule.

**Comment by hvy** [comment]
Dumping some plots from short runs. All Lookahead hyperparameters are set to default. Just for the record, learning rates of the base optimizer should probably be increased, by roughly a factor of 10 or so when using Lookahead. This requires more runs to be verified though.

- ResNet50 (chainercv), CIFAR-100, `CorrectedMomentumSGD` (default params but `lr=0.15`)
  - w/o Lookahead
loss_no_lookahead_no_exp_shift
  - w/ Lookahead
loss_w_lookahead_no_exp_shift

- ResNet50 (chainercv), CIFAR-10, `MomentumSGD` (default params), `ExponentialShift` (see [train_cifar.py](https://github.com/chainer/chainer/blob/master/examples/cifar/train_cifar.py))
  - w/o Lookahead
loss_wo_lookahead
  - w/ Lookahead
loss_w_lookahead
_Exponential shift reduced the learning rate of the alpha in Lookahead, and thus slowed down training. This was a misconfiguration._
